plugin/restore/mysql/mysql.go

```diff
@@ -457,92 +492,193 @@ func getPITROldDatabaseName(database string, suffixTs int64) string {
+	}
+	return binlogFilesLocalSorted, nil
+}
 
-	// build a local file size map from file name to size
-	localFileMap := make(map[string]int64)
+func localBinlogFilesAreContinuous(files []BinlogFile) bool {
+	for i := 0; i < len(files)-1; i++ {
+		if files[i].Seq+1 != files[i+1].Seq {
+			return false
+		}
+	}
+	return true
+}
 
-	for _, localFile := range binlogFilesLocal {
-		localFileMap[localFile.Name()] = localFile.Size()
+// FetchBinlogFilesUpToTargetTs downloads the locally missing binlog files required to replay to `targetTs` from the remote instance to `binlogDir`.
+// After downloading a binlog file, we check its first event timestamp. If the timestamp is larger than
+// targetTs, we stop the following downloads and return eagerly, because now the local binlog files contain
+// all the binlog events we need for this recovery task.
+func (r *Restore) FetchBinlogFilesUpToTargetTs(ctx context.Context, targetTs int64) error {
+	// Read binlog files list on server.
+	binlogFilesOnServerSorted, err := r.GetSortedBinlogFilesMetaOnServer(ctx)
+	if err != nil {
+		return err
+	}
+	if len(binlogFilesOnServerSorted) == 0 {
+		// No binlog files on server, so there's nothing to download.
+		log.Debug("No binlog file found on server")
+		return nil
 	}
+	latestBinlogFileOnServer := binlogFilesOnServerSorted[len(binlogFilesOnServerSorted)-1]
+	log.Debug("Got sorted binlog file list on server", zap.Array("list", ZapBinlogFiles(binlogFilesOnServerSorted)))
 
-	todo := make(map[string]bool)
+	// Read the local binlog files.
+	binlogFilesLocalSorted, err := getLocalBinlogFiles(r.binlogDir)
+	if err != nil {
+		return fmt.Errorf("failed to read local binlog files, error[%w]", err)
+	}
 
-	for _, serverBinlog := range binlogFilesOnServer {
-		// We don't download the latest binlog in SyncArchivedBinlogFiles()
-		if serverBinlog.Name == latestBinlogFileOnServer.Name {
-			continue
+	if len(binlogFilesLocalSorted) != 0 {
+		// Re-download corrupted local binlog files.
+		// This is a best-effort task, because the corresponding binlog files may not exist on server.
+		log.Debug("Re-downloading corrupted local binlog files")
+		binlogFilesOnServerMap := make(map[string]BinlogFile)
+		for _, file := range binlogFilesOnServerSorted {
+			binlogFilesOnServerMap[file.Name] = file
 		}
-
-		localBinlogSize, ok := localFileMap[serverBinlog.Name]
-		if !ok {
-			todo[serverBinlog.Name] = true
-			continue
+		for _, file := range binlogFilesLocalSorted {
+			fileOnServer, existOnServer := binlogFilesOnServerMap[file.Name]
+			if existOnServer && file.Size != fileOnServer.Size {
+				path := filepath.Join(r.binlogDir, file.Name)
+				log.Debug("Deleting corrupted local binlog file",
+					zap.String("path", path),
+					zap.Int64("sizeLocal", file.Size),
+					zap.Int64("sizeOnServer", fileOnServer.Size))
+				if err := os.Remove(path); err != nil {
+					log.Error("Failed to remove corrupted local binlog file", zap.String("path", path), zap.Error(err))
```

Comment:
inconsistent instead of corrupted? This happens like to the last binlog file which is not caused by corruption.

Reply:
Done.

Reply:
Still corrupted

Reply:
Oops... Just pushed...